stustapay/tse/diebold_nixdorf_usb/tse_verifier.py

```diff
@@ -0,0 +1,108 @@
+import ecdsa
+from hashlib import sha384, sha256
+import base64
+from dateutil import parser
+from datetime import timezone
+import pprint
+from asn1crypto.core import Integer, ObjectIdentifier, Sequence, OctetString, PrintableString, Any
+import argparse
+
+
+# initialize data structures
+class SignatureAlgorithm_seq(Sequence):
+    _fields = [
+        ("signatureAlgorithm", ObjectIdentifier),
+        ("parameters", OctetString, {"optional": True}),
+    ]
+
+
+class CertifiedData(Sequence):
+    _fields = [
+        ("operationType", PrintableString, {"implicit": 0}),
+        ("clientId", PrintableString, {"implicit": 1}),
+        ("ProcessData", OctetString, {"implicit": 2}),
+        ("ProcessType", PrintableString, {"implicit": 3}),
+        ("additionalExternalData", OctetString, {"implicit": 4, "optional": True}),
+        ("transactionNumber", Integer, {"implicit": 5}),
+        ("additionalInternalData", OctetString, {"implicit": 6, "optional": True}),
+    ]
+
+
+class TransactionData(Sequence):
+    _fields = [
+        ("Version", Integer),
+        ("CertifiedDataType", ObjectIdentifier),
+        ("CertifiedData", Any),
+        ("SerialNumber", OctetString),
+        ("signatureAlgorithm", Sequence),
+        ("signatureCounter", Integer),
+        ("LogTime", Integer),  # unixTime!
+    ]
+
+
+def main(args):
+    qr = args.QR.split(";")
+
+    certidata = CertifiedData()
+    data = TransactionData()
+
+    if qr[0] != "V0":  # check version
+        raise NotImplemented
+
+    certidata["clientId"] = qr[1]  # fill in data fields from QR code in prepared ASN.1 structure
+    certidata["ProcessType"] = qr[2]
+    certidata["ProcessData"] = qr[3].encode("utf-8")
+    certidata["transactionNumber"] = int(qr[4])
+    data["signatureCounter"] = int(qr[5])
+    data["LogTime"] = int(parser.isoparse(qr[7].split(".")[0]).replace(tzinfo=timezone.utc).timestamp())
+
+    public_key = qr[11]
+    if qr[8] != "ecdsa-plain-SHA384":
+        raise NotImplemented
+    else:
+        signaturealgorithm = SignatureAlgorithm_seq()
+        signaturealgorithm["signatureAlgorithm"] = "0.4.0.127.0.7.1.1.4.1.4"  # ecdsa-plain-SHA384
+        vk = ecdsa.VerifyingKey.from_string(base64.b64decode(public_key), curve=ecdsa.BRAINPOOLP384r1, hashfunc=sha384)
+
+    if qr[9] != "unixTime":  # case sensitive != UnixTime ...
```

Comment:
what's the difference :D

Reply:
'u' != 'U'

Reply:
qrcode has to have `unixTime`, but the DN export returns `UnixTime`...